curp/src/server/storage/wal/codec.rs

```diff
@@ -0,0 +1,393 @@
+            if let Some((frame, len)) = WALFrame::<C>::decode(src)? {
+                let decoded_bytes = src.split_to(len);
+                match frame {
+                    WALFrame::Data(data) => {
+                        self.frames.push(data);
+                        self.hasher.update(decoded_bytes);
+                    }
+                    WALFrame::Commit(commit) => {
+                        let frames_bytes: Vec<_> =
+                            self.frames.iter().flat_map(DataFrame::encode).collect();
+                        let checksum = self.hasher.clone().finalize();
+                        self.hasher.reset();
+                        if commit.validate(&checksum) {
+                            return Ok(Some(self.frames.drain(..).collect()));
+                        }
+                        return Err(WALError::Corrupted(CorruptType::Checksum));
+                    }
+                }
+            } else {
+                return Ok(None);
+            }
+        }
+    }
+}
+
+#[allow(
+    clippy::indexing_slicing, // Index slicings are checked
+    clippy::integer_arithmetic, //  Arithmetics are checked
+    clippy::unnecessary_wraps // Use the wraps to make code more consistenct
+)]
+impl<C> WALFrame<C>
+where
+    C: for<'a> Deserialize<'a>,
+{
+    /// Decodes a frame from the buffer
+    ///
+    /// * The frame header memory layout
+    ///
+    /// 0      1      2      3      4      5      6      7      8
+    /// |------+------+------+------+------+------+------+------|
+    /// | Type | Length / Index / Reserved                      |
+    /// |------+------+------+------+------+------+------+------|
+    ///
+    /// * The frame types
+    ///
+    /// |------------+-------+-------------------------------------------------------|
+    /// | Type       | Value | Desc                                                  |
+    /// |------------+-------+-------------------------------------------------------|
+    /// | Invalid    |  0x00 | Invalid type                                          |
+    /// | Entry      |  0x01 | Stores record of CURP log entry                       |
+    /// | Seal Index |  0x02 | Stores the highest index of this current sealed frame |
+    /// | Commit     |  0x03 | Stores the checksum                                   |
+    /// |------------+-------+-------------------------------------------------------|
+    fn decode(src: &[u8]) -> Result<Option<(Self, usize)>, WALError> {
+        if src.len() < 8 {
+            return Ok(None);
+        }
+        let header: [u8; 8] = src[0..8]
+            .try_into()
+            .unwrap_or_else(|_| unreachable!("this conversion will always succeed"));
+        let frame_type = header[0];
+        match frame_type {
+            0x00 => Err(WALError::MaybeEnded),
+            0x01 => Self::decode_entry(header, &src[8..]),
+            0x02 => Self::decode_seal_index(header),
+            0x03 => Self::decode_commit(&src[8..]),
```

Comment:
We can use a set of constants to represent `frame_type` to enhance readability

Reply:
done